apiserver/src/main/resources/migration/changelog-v5.6.0.xml

```diff
@@ -1430,4 +1430,282 @@
+            ) PARTITION BY RANGE ("LAST_OCCURRENCE");
+
+            -- create new index
+            CREATE INDEX "PORTFOLIOMETRICS_NEW_LAST_OCCURRENCE_IDX" ON "PORTFOLIOMETRICS_NEW" ("LAST_OCCURRENCE");
+
+            -- call the migration procedure
+            CALL "MIGRATE_METRICS_TO_PARTITIONS"('PORTFOLIOMETRICS_NEW'::TEXT, 'PORTFOLIOMETRICS'::TEXT);
+
+            -- drop old table
+            DROP TABLE "PORTFOLIOMETRICS";
+
+            -- rename partitioned table
+            ALTER TABLE "PORTFOLIOMETRICS_NEW" RENAME TO "PORTFOLIOMETRICS";
+
+            -- rename PK
+            ALTER TABLE "PORTFOLIOMETRICS" RENAME CONSTRAINT "PORTFOLIOMETRICS_NEW_PK" TO "PORTFOLIOMETRICS_PK";
+            ALTER SEQUENCE "PORTFOLIOMETRICS_NEW_ID_seq" RENAME TO "PORTFOLIOMETRICS_ID_seq";
+
+            -- rename index
+            ALTER INDEX "PORTFOLIOMETRICS_NEW_LAST_OCCURRENCE_IDX" RENAME TO "PORTFOLIOMETRICS_LAST_OCCURRENCE_IDX";
+        </sql>
+        <!--Migration for "PROJECTMETRICS"-->
+        <sql>
+            -- create new partitioned table with new PK
+            CREATE TABLE "PROJECTMETRICS_NEW"
+            (
+                "ID" BIGSERIAL,
+                "COMPONENTS" int4 NOT NULL,
+                "CRITICAL" int4 NOT NULL,
+                "FINDINGS_AUDITED" int4 NULL,
+                "FINDINGS_TOTAL" int4 NULL,
+                "FINDINGS_UNAUDITED" int4 NULL,
+                "FIRST_OCCURRENCE" timestamptz NOT NULL,
+                "HIGH" int4 NOT NULL,
+                "RISKSCORE" float8 NOT NULL,
+                "LAST_OCCURRENCE" timestamptz NOT NULL,
+                "LOW" int4 NOT NULL,
+                "MEDIUM" int4 NOT NULL,
+                "POLICYVIOLATIONS_AUDITED" int4 NULL,
+                "POLICYVIOLATIONS_FAIL" int4 NULL,
+                "POLICYVIOLATIONS_INFO" int4 NULL,
+                "POLICYVIOLATIONS_LICENSE_AUDITED" int4 NULL,
+                "POLICYVIOLATIONS_LICENSE_TOTAL" int4 NULL,
+                "POLICYVIOLATIONS_LICENSE_UNAUDITED" int4 NULL,
+                "POLICYVIOLATIONS_OPERATIONAL_AUDITED" int4 NULL,
+                "POLICYVIOLATIONS_OPERATIONAL_TOTAL" int4 NULL,
+                "POLICYVIOLATIONS_OPERATIONAL_UNAUDITED" int4 NULL,
+                "POLICYVIOLATIONS_SECURITY_AUDITED" int4 NULL,
+                "POLICYVIOLATIONS_SECURITY_TOTAL" int4 NULL,
+                "POLICYVIOLATIONS_SECURITY_UNAUDITED" int4 NULL,
+                "POLICYVIOLATIONS_TOTAL" int4 NULL,
+                "POLICYVIOLATIONS_UNAUDITED" int4 NULL,
+                "POLICYVIOLATIONS_WARN" int4 NULL,
+                "PROJECT_ID" int8 NOT NULL,
+                "SUPPRESSED" int4 NOT NULL,
+                "UNASSIGNED_SEVERITY" int4 NULL,
+                "VULNERABILITIES" int4 NOT NULL,
+                "VULNERABLECOMPONENTS" int4 NOT NULL,
+                CONSTRAINT "PROJECTMETRICS_NEW_PK" PRIMARY KEY ("ID", "LAST_OCCURRENCE")
```

Comment:
We may want to use `PROJECT_ID, LAST_OCCURRENCE` as primary key instead:

* Our lookups are almost always `where project_id = ? and last_occurrence >= ?`, so we'd need an index on both columns anyway. Primary keys create a unique index behind the scenes already.
* We want to enforce uniqueness such that there can be no duplicate records for the same `PROJECT_ID`<->`LAST_OCCURRENCE` combination. A primary key achieves that.

Following this, the `ID` column would be no longer needed.

For `PORTFOLIOMETRICS`, `LAST_OCCURRENCE` alone can form a primary key.

Reply:
And for `DEPENDENCYMETRICS` it should be `PROJECT_ID, COMPONENT_ID, LAST_OCCURRENCE` or just `COMPONENT_ID, LAST_OCCURRENCE`

Reply:
Good question. We *do* query these by `PROJECT_ID` sometimes, for example when updating project metrics: https://github.com/DependencyTrack/hyades-apiserver/blob/dc9a1bf2e27374c6bb553c92fd281c0aac597a6e/apiserver/src/main/resources/migration/procedures/procedure_update-project-metrics.sql#L80-L83

So I think `PROJECT_ID, COMPONENT_ID, LAST_OCCURRENCE` would make sense.

Reply:
I'll have to keep the column "ID" though because if removed, migration is throwing exception `ERROR: INSERT has more expressions than target columns`

```
-- Insert data from existing table into this partition
                    EXECUTE format(
                        'INSERT INTO %I SELECT * FROM %I WHERE "LAST_OCCURRENCE" >= %L AND "LAST_OCCURRENCE" < %L;',
                        target_table,
                        source_table,
                        partition_date,
                        next_day
                    );
```

Reply:
I think this is "just" a matter of expanding `SELECT *` to all the columns we want (i.e. all except `ID`).

Reply:
I know I was avoiding the big query :D I'll edit it 👍🏼

Reply:
Two options I can think of:

* Define variables for each table with columns to select, then include them in the query via `%s`.
* Keep the `ID` column for the migration, but then immediately drop it afterwards. The downside is that dropping a column [doesn't really remove it](https://stackoverflow.com/a/15700185).

I would prefer the former. It's not pretty, but then again these procedures are temporary so they don't have to be.

WDYT? Any other options you see?

Reply:
I see online that the disk space for dropped column in Postgres can be reclaimed using [VACUUM](https://www.postgresql.org/docs/current/sql-vacuum.html) or [CLUSTER](https://www.postgresql.org/docs/current/sql-cluster.html) but it may slow it down if there is lots of migrated data in the table because whole table will get locked for it.

Reply:
To avoid any possible downtime, list of columns is the best way to do it. 👍🏼 Pushed the changes.